### DISCUSSION TITLE
How to use NumericSequenceGenerator on a type's property

### DISCUSSION THREAD

**Discussion by Kralizek** [AutoFixture/AutoFixture#1192] [Q&A]
Hi,

I have a small type like the following one:

```csharp
public class Contact
{
	public int Id { get; set; }
	
	public string FirstName { get; set; }

	public string LastName { get; set; }

	public int Age { get; set; }
}
```

I'd like to configure my fixture to use the `NumericSequenceGenerator` on the `Id` property.

I know that when customizing a type, `FromFactory` accepts a `ISpecimenBuilder` but this is not the same for `With`.

Ideally, I'd like to do something like the snippet below, but I don't think it's supported.

```csharp
fixture.Customize<Contact>(c => c
     .With(p => p.Id, new NumericSequenceGenerator())
);
```

I even tried with using the delegate overload

```csharp
fixture.Customize<Contact>(
	c => c.With(p => p.Id, (NumericSequenceGenerator generator) => generator.Create<int>())
);
```

but it throws an exception:

```
ObjectCreationExceptionWithPath: AutoFixture was unable to create an instance from UserQuery+Contact because creation unexpectedly failed with exception. Please refer to the inner exception to investigate the root cause of the failure.

Request path:
	UserQuery+Contact

Inner exception messages:
	System.InvalidCastException: Unable to cast object of type 'AutoFixture.Kernel.NoSpecimen' to type 'System.Int32'.
```

Right now I'm using
```csharp
fixture.Customize<int>(c => c.FromFactory(new NumericSequenceGenerator()));
```
which is equivalent to
```csharp
fixture.Customizations.Add(new NumericSequenceGenerator());
```

The issue with this approach is that every integer is affected

Id | FirstName | LastName | Age
-- | -- | -- | --
1 | FirstName421ad33a-3df9-4b15-ab54-093220411499 | LastName5854d3ab-25c7-420a-9c70-7a93d825f720 | 2
3 | FirstNameff6a888f-19f7-4838-9dbf-8f2a7db841cb | LastName4f7ef66f-72ee-409e-a5b7-32a682b83dc5 | 4
5 | FirstNamec5cb2b12-5054-4b2b-b0ec-f823c8b9ca3a | LastName27f43eb8-a584-4e9e-b1bf-f4e655d962cb | 6

**Comment by Ergamon** [comment]
Seems you are hunting the wrong problem ;)

The extension method Create<T> on NumericGenerator does not work like you think.

Just write the following code and you will see the excpetion:
```
var generator = new NumericSequenceGenerator();
int number = generator.Create<int>();
```
It looks a littly ugly, but if you write it this way, it works:
```
var generator = new NumericSequenceGenerator();
int number = (int)generator.Create(typeof(int), null);
```
In the end I may ask the question why to use this generator at all? I mean if you want to limit it to a single property of a single type you can do the following:
```
var fixture = new Fixture();

int sequenceNumber = 1;
fixture.Customize<Contact>(c => c.With(contact => contact.Id, () => sequenceNumber++));

var contacts = fixture.CreateMany<Contact>(10);

foreach (var contact in contacts)
{
  Console.WriteLine($"{contact.Id,2}: {contact.FirstName} {contact.LastName} ({contact.Age,3})");
}
```
So the With extension method works fine, as you can see.

**Comment by Kralizek** [comment]
Hi @Ergamon,

thanks for your reply. Your snippet works as expected. Thanks.

In reality, the question was meant to be a bit more generic because there are many cases where you want to use a specific `ISpecimenBuilder` for just one property without affecting the global setup for the type.

I suggest looking into the possibility of creating an overload of `With` that accepts a `ISpecimenBuilder`. Something like `With<TProperty>(Expression<Func<T, TProperty>> propertyPicker, ISpecimenBuilder builder)` to be used like below

```csharp
fixture.Customize<Contact>(c => c
    .With(p => p.Id, new NumericSequenceGenerator())
    .With(p => p.DateOfBirth, new RandomDateTimeSequenceGenerator(new DateTime(1980, 1, 1), new DateTime(1985, 12, 31)))
);
```

**Comment by Ergamon** [comment]
As we both know, development here is not too heavy ;)

I looked into the source and the problem seems to be that the Create<T> extension generates a SeededRequest<T> request and not a typeof(T) request. NumericSequenceGenerator cannot handle SeededRequest<T>.

If you want to have your syntax you can just use an extension like this:
```
  public static class PostprocessComposerExtensions
  {
    public static IPostprocessComposer<T> With<T, TProperty>(
      this IPostprocessComposer<T> postprocessComposer,
      Expression<Func<T, TProperty>> propertyPicker,
      ISpecimenBuilder specimenBuilder)
    {
      return postprocessComposer
        .With(propertyPicker, () => specimenBuilder.Create<TProperty>());
    }

    private static T Create<T>(this ISpecimenBuilder specimenBuilder)
    {
      var request = typeof(T);
      var context = new SpecimenContext(specimenBuilder);
      var specimen = specimenBuilder.Create(request, context);

      if (specimen is NoSpecimen)
      {
        var specimenBuilderType = specimenBuilder.GetType();
        var message = $"Specimen builder {specimenBuilderType} could not generate type {request}.";
        throw new InvalidOperationException(message);
      }

      return (T)specimen;
    }
  }
```

**Comment by Kralizek** [comment]
oh wow! thanks @Ergamon It worked like a charm. Maybe you could consider packaging it as a PR and let @aivascu decide if it's something to be merged

**Comment by aivascu** [comment]
@Kralizek this issue reminds me of a similar suggestion #1180, to simplify the specification of `SpecimenBuilders` for particular types. Perhaps we indeed need more syntactic sugar around the builders.

As for the specific extensions proposed by @Ergamon, I see an issue in the sense that not every specimen builder accepts the `Type` as a request object. An example would be the `RegularExpressionGenerator` or the `RangedNumberGenerator`.
These builders require specific parameterized requests to be sent, which cannot be inferred from the type of the request alone.

A naive solution, I can suggest is making use of the factory method overload of the `With` extension and provide it a factory containing only the strictly necessary context to complete.
```csharp
var fixture = new Fixture();
fixture.Customize<PropertyHolder>(
    x => x.With(x => x.MyProperty, NumericFactory.Create<int>));
/* ... */
public static class NumericFactory
{
    public static T Create<T>()
        where T : struct, IComparable<T>, IConvertible
    {
        var specimen = new NumericSequenceGenerator().Create(typeof(T), null);

        if (specimen is T value) return value;

        throw new InvalidOperationException(
            $"Specimen builder {typeof(NumericSequenceGenerator)} could not generate type {typeof(T)}.");
    }
}
```
this would allow you to customize the request before it is processed by the specimen builder.

Of course this is not ideal as well since some builders require more "context". For example the same `RangedNumberGenerator` during the creation requests an instance of a nubmer from the context. So this implementation would require more setup than just passing a single builder.

```csharp
public static T FromRange<T>(T min, T max)
    where T : struct, IComparable<T>, IConvertible
{
    var specimen = new RangedNumberGenerator()
        .Create(
            new RangedNumberRequest(typeof(T), min, max),
            new SpecimenContext(new NumericSequenceGenerator()));

    if (specimen is T value) return value;

    throw new InvalidOperationException(
        $"Specimen builder {typeof(NumericSequenceGenerator)} could not generate type {typeof(T)}.");
}
```
Perhaps there is a better way to approach this problem this is what came up from the top of my head.

**Comment by Kralizek** [comment]
@aivascu 
Maybe by introducing abstract classes implementing the `ISpecimenBuilder` interface to distinguish builders by their accepted request type? This would allow us to distinguish a bit better between builders.

**Comment by aivascu** [comment]
@Kralizek could you provide an example? I don't think I understand your idea.

**Comment by Kralizek** [comment]
@aivascu

The problem, as I see it, is that `ISpecimenBuilder` is a very wide interface exposing a method whose parameters are extremely wide to the point that we can't make any assumption on what we get or need, some would even go as far as seeing it as a violation of _LSP_.
 
I was thinking about introducing parent classes like the one below and modify the existing builders to inherit from these classes. 

```csharp
public abstract class TypeSpecimenBuilder : ISpecimenBuilder
{
	public object Create(object request, ISpecimenContext context)
	{
		if (request is Type type)
		{
			return Create(type, context);
		}
		
		return new NoSpecimen();
	}
	
	public abstract object Create(Type request, ISpecimenContext context);
}

public class NumericSequenceGenerator : TypeSpecimenBuilder, ISpecimenBuilder
{
	public virtual object Create(Type request, ISpecimenContext context)
	{
		return this.CreateNumericSpecimen(request);
	}

	...
}
```

This would allow us to distinguish between families of builders and leverage this knowledge to empower the API like below

```csharp
    public static IPostprocessComposer<T> With<T, TProperty>(
      this IPostprocessComposer<T> postprocessComposer,
      Expression<Func<T, TProperty>> propertyPicker,
      TypeSpecimenBuilder specimenBuilder)
    {
      return postprocessComposer
        .With(propertyPicker, specimenBuilder.Create(typeof(TProperty),null));
    }
```

**Comment by aivascu** [comment]
@Kralizek I see what you mean and I think this solution might work for a certain number of specimen builders.
However the SpecimenBuilders are not usually as simple. It is common that the create method can return a valid response whether the request is a `Type` or a `PropertyInfo`.

A more generic and probably less intrusive solution could be to customize the properties using a "context-aware" factory method, that would compose a request from any given parameters, either just the type or some additional values, like for example a range of numbers and then just pass the request object to the current context and let the context (fixture) figure out how to resolve the request.

```csharp
fixture.Customize<Customer>(
   x => x.With(c => c.Age, s => s.FromRange(18, 89))
         .With(c => c.FirstName, s => s.FirstName(locale: "en-GB"))
         .With(c => c.Email, s => s.EmailAddress(domain: "example.com"))
);
```
I think that this kind API would be much more expressive and extensible since all the methods on the right-hand side could be extension methods that compose request objects.

Same could apply for type builders.
```csharp
fixture.Build<Customer>()
       .With(c => c.Age, s => s.FromRange(22, 100))
       .With(c => c.FirstName, s => s.FirstName(locale: "de-DE"))
       .With(c => c.Email, s => s.EmailAddress(domain: "foo.co.uk"))
       .Create();
```

**Comment by Kralizek** [comment]
Interesting idea! @aivascu 

Could you sketch down the implementation of one of these extension methods?

**Comment by aivascu** [comment]
@Kralizek sorry for the delay. As it turns out the behavior I've described is already possible using for example value factory override of the `IPostProcessComposer`.
```csharp
[Fact]
public void BuildShouldInvoke()
{
    var fixture = new Fixture();

    var holder = fixture.Build<PropertyHolder<int>>()
        .With<int, IFixture>(x => x.Property, (x) => x.FromRange(538, 827))
        .CreateMany(15);

    Assert.True(holder.All(x => x.Property >= 538 && x.Property <= 827));
}

public static class RequestExtensions
{
    public static T FromRange<T>(this ISpecimenBuilder builder, T min, T max)
    {
        return (T)builder.Create(
            new RangedNumberRequest(typeof(T), min, max),
            new SpecimenContext(builder));
    }
}
```
Assuming you have customized beforehand the `Fixture` instance, with the appropriate customization, you can resolve any complex request like the ones proposed above.

By adding an overload for the `With` method we can get rid of the explicit type specification and obtain a syntax like this.
```csharp
[Fact]
public void BuildShouldInvoke1()
{
    var fixture = new Fixture();

    var holder = fixture.Build<PropertyHolder<int>>()
        .With(x => x.Property, (x) => x.FromRange(15, 20))
        .CreateMany(15);

    Assert.True(holder.All(x => x.Property >= 15 && x.Property <= 20));
}
```
